### PR TITLE
[SPARK-47475][CORE][K8S] Support `spark.kubernetes.jars.avoidDownloadSchemes` for K8s Cluster Mode

### PR DESCRIPTION
**Pull request by leletan**
### What changes were proposed in this pull request?

During spark submit, for K8s cluster mode driver, instead of always downloading jars and serving it to executors, avoid the download if the url matches `spark.kubernetes.jars.avoidDownloadSchemes` in the configuration.

### Why are the changes needed?

For K8s cluster mode driver, `SparkSubmit` will download all the jars in the `spark.jars` to driver and then those jars' urls in `spark.jars` will be replaced by the driver local paths. Later when driver starts the `SparkContext`, it will copy all the `spark.jars` to `spark.app.initial.jar.urls`, start a file server and replace the jars with driver local paths in `spark.app.initial.jar.urls` with file service urls. When the executors start, they will download those driver local jars by `spark.app.initial.jar.urls`.
When jars are big and the spark application requests a lot of executors, the executors' massive concurrent download of the jars from the driver will cause network saturation. In this case, the executors jar download will timeout, causing executors to be terminated. From user point of view, the application is trapped in the loop of massive executor loss and re-provision, but never gets enough live executors as requested, leads to SLA breach or sometimes failure. 
So instead of letting driver to download the jars and then serve them to executors, if we just avoid driver from downloading the jars and keeping the urls in `spark.jars` as they were, the executor will try to directly download the jars from the urls provided by user. This will avoid the driver download bottleneck mentioned above, especially when jar urls are with scalable storage schemes, like s3 or hdfs.
Meanwhile, there are cases jar urls are with schemes of less scalable than driver file server, e.g. http, ftp, etc, or when the jars are small, or executor count is small - user may still want to fall back to current solution and use driver file server to serve the jars.
So in this case, make the driver jars downloading and serving optional by scheme (similar idea to `FORCE_DOWNLOAD_SCHEMES` in YARN) is a good approach for the solution.

### Does this PR introduce _any_ user-facing change?

A configuration `spark.kubernetes.jars.avoidDownloadSchemes` is added

### How was this patch tested?

- Unit tests added
- Tested with an application running on AWS EKS submitted with a 1GB jar on s3.
  - Before the fix, the application could not scale to 1k live executors. 
  - After the fix, the application had no problem to scale beyond 12k live executors.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -401,16 +401,24 @@ private[spark] class SparkSubmit extends Logging {
         // SPARK-33782 : This downloads all the files , jars , archiveFiles and pyfiles to current
         // working directory
         // SPARK-43540: add current working directory into driver classpath
+        // SPARK-47475: make download to driver optional so executors may fetch resource from remote
+        // url directly to avoid overwhelming driver network when resource is big and executor count
+        // is high
         val workingDirectory = "."
         childClasspath += workingDirectory
-        def downloadResourcesToCurrentDirectory(uris: String, isArchive: Boolean = false):
+
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:409]: We don't need to add a new empty line here.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -401,16 +401,24 @@ private[spark] class SparkSubmit extends Logging {
...
+        def downloadResourcesToCurrentDirectory(uris: String,
+                                                isArchive: Boolean = false,
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:411]: Could you take a look at `Apache Spark Coding Guideline` at this time?
- https://spark.apache.org/contributing.html
> For Scala code, Apache Spark follows the official [Scala style guide](http://docs.scala-lang.org/style/) and [Databricks Scala guide](https://github.com/databricks/scala-style-guide). The latter is preferred. To format Scala code, run ./dev/scalafmt prior to submitting a PR.

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:411]: For this case, the following is the rule.
> For method declarations, use 4 space indentation for their parameters

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a narrow-downed PR. I did a quick review today.  I'll take another round review tomorrow, @leletan .

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 3]
```diff
@@ -1458,6 +1458,18 @@ package object config {
       .doubleConf
       .createWithDefault(1.5)
 
+  private[spark] val KUBERNETES_AVOID_JAR_DOWNLOAD_SCHEMES =
+    ConfigBuilder("spark.kubernetes.jars.avoidDownloadSchemes")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1462]: BTW, the AS-IS PR title match with this?
> Make Jars Download to Driver Optional under K8s Cluster Mode

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 2 above)
Comment (leletan) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:411]: Ran ` ./dev/scalafmt` as mentioned, it actually modified a lot of existing code style as well - is this expected? Or something I missed here?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:411]: Ya, `scalafmt` is only recommended for newly added code. Please don't touch the existing code.

Comment (leletan) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:411]: Thanks for the tips @dongjoon-hyun! 
I tried running `./dev/scalafmt`, somehow a lot of existing code was refactored for style. Is this expected?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @leletan .

**Pull request closed without merging**